Move non-tensor nodes on boundary after split

openzig opened the pull request:
Summary: See details in the attached doc
[AOTI Non-Tensor on Boundary Issue.pdf](https://github.com/user-attachments/files/22608882/AOTI.Non-Tensor.on.Boundary.Issue.pdf)

Meta internal post: https://fb.workplace.com/groups/aoti.productionization/permalink/743060011864813/


Test Plan: unit test

Bifferential Revision: D80903201

Steps to show how this feature works:
- checkout commit: 01eb5dd5c3dc3e5d62fc975da9bcfee4befbd421
- run this local lowering command: P2072906967
You will see error like below, this means there is a node on the boundary whose dtype is `torch.dtype`, the corresponding model code is [this line](https://www.internalfb.com/code/fbsource/[e9ae35402388]/fbcode/minimal_viable_ai/umia_v1/ig/omni/root_model.py?lines=2409)
```
Found <class 'torch.dtype'> in output, which is not a known type.
```
- Then search `move_non_tensor_nodes_on_boundary` in file `fbcode/caffe2/torch/fb/model_transform/lower_presets/aoti/ig_ss_t2i_retrieval_aot_inductor.py`, you will see it is commented, now set `move_non_tensor_nodes_on_boundary` to true, and rerun the command in last step, you will see the tracing is done and `_exported_run_on_acc_1_readable.txt` is generated. (although there is IMA error which is not related to this diff)


cc @ezyang @EikanWang @jgong5 @wenzhe-nrv

sevenEng commented on torch/fx/passes/splitter_base.py:
```diff
@@ -119,6 +120,16 @@ def __init__(
+            required=False,
+            type=bool,
+            help="AOTI does not support non-tensor nodes on acc->acc, acc->gpu and gpu->acc boundary. "
+            "For non-tensor nodes on acc->acc boundary and acc->gpu, we move the nodes from upstream to downstream. "
```
i'm thinking will acc->gpu also already handled by `reduce_acc_nodes_non_tensor_output`

inside `reduce_acc_nodes_non_tensor_output`, we'll find these nodes who are in acc but produce non-tensor output and have users in non-acc, and then move them out of acc

sevenEng commented on torch/fx/passes/splitter_base.py:
```diff
@@ -1092,6 +1104,8 @@ def split(self, remove_tag: bool = False) -> torch.fx.GraphModule:
         subgraphs = self.put_nodes_into_subgraphs()
         subgraphs = self.remove_small_acc_subgraphs(subgraphs)
+        if self.settings.move_non_tensor_nodes_on_boundary:
+            move_non_tensor_nodes_on_boundary(subgraphs)
```
i'm wondering if it's possible to integrate this with the node tracker in some way,

as here we're changing the split result, and the node movements we tracked in it will actually be wrong

```
        self.tracker = NodeEventTracker(int(TRACKER_MODE), DUMP_PREFIX)
```

ColinPeppler replied:
Is there a good reason to call `move_non_tensor_nodes_on_boundary` after `remove_small_acc_subgraphs`? I think it would make more sense to call it before because it can modify the acc subgraphs (e.g. non-tensor node move causes acc subgraph to shrink).

ColinPeppler commented on torch/fx/passes/split_utils.py:
```diff
@@ -310,3 +316,192 @@ def remap_func(x):
+    1. Find nodes whose type is not tensor and any of its children is in another
+       subgraph, put them in a queue for next step
+
+    2. Do a BFS on those nodes, for each node, let's say node X and it is in subgraph A:
```
I'm thinking you're referring to `can_move_node_and_dependencies` here.

```suggestion
    2. Do a DFS on those nodes, for each node, let's say node X and it is in subgraph A:
```

ColinPeppler commented on torch/fx/passes/splitter_base.py:
```diff
@@ -120,6 +119,16 @@ def __init__(
+            "--move-non-tensor-nodes-on-boundary",
+            "--move_non_tensor_nodes_on_boundary",
+            required=False,
+            type=bool,
```
From [argparse docs](https://docs.python.org/3/library/argparse.html#:~:text=The%20bool()%20function%20is%20not%20recommended%20as%20a%20type%20converter.%20All%20it%20does%20is%20convert%20empty%20strings%20to%20False%20and%20non%2Dempty%20strings%20to%20True.%20This%20is%20usually%20not%20what%20is%20desired.)

> The [bool()](https://docs.python.org/3/library/functions.html#bool) function is not recommended as a type converter. All it does is convert empty strings to False and non-empty strings to True. This is usually not what is desired.

```suggestion
            action='store_true',
```

ColinPeppler reviewed (approved):
lgtm with 2 follow ups

openzig commented:
@pytorchmergebot merge

openzig committed:
Move non-tensor nodes on boundary after split (#163605)

Summary:

See details in the post: https://fb.workplace.com/groups/aoti.productionization/permalink/743060011864813/ 

For external users, please see details in attached file
{F1982376254}

Steps to show how this feature works:
- checkout commit: 01eb5dd5c3dc3e5d62fc975da9bcfee4befbd421
- run this local lowering command: P2072906967
You will see error like below, this means there is a node on the boundary whose dtype is `torch.dtype`, the corresponding model code is [this line](https://www.internalfb.com/code/fbsource/[e9ae35402388]/fbcode/minimal_viable_ai/umia_v1/ig/omni/root_model.py?lines=2409)
```
Found <class 'torch.dtype'> in output, which is not a known type.
```
- Then search `move_non_tensor_nodes_on_boundary` in file `fbcode/caffe2/torch/fb/model_transform/lower_presets/aoti/ig_ss_t2i_retrieval_aot_inductor.py`, you will see it is commented, now set `move_non_tensor_nodes_on_boundary` to true, and rerun the command in last step, you will see the tracing is done and `_exported_run_on_acc_1_readable.txt` is generated. (although there is IMA error which is not related to this diff)

Test Plan: unit test

Reviewed By: sevenEng, xunzhang, ColinPeppler

Differential Revision: D80903201
